Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean

```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
+    (h : WalkingReflexivePairHom Z W) : comp (comp f g) h = comp f (comp g h) := by
+  cases f <;> cases g <;> cases h <;> rfl
+
+instance walkingReflexivePairCategory : SmallCategory WalkingReflexivePair where
+  Hom := WalkingReflexivePairHom
+  id := id
+  comp := comp
+  comp_id := comp_id
+  id_comp := id_comp
+  assoc := assoc
+
+@[simp]
+theorem walkingReflexivePairHom_id (X : WalkingReflexivePair) :
+    WalkingReflexivePairHom.id X = 𝟙 X := by rfl
+
+@[simp]
+theorem walkingReflexivePairHom_left_comp_reflexion : left.comp reflexion = left_reflexion := rfl
+
+@[simp]
+theorem walkingReflexivePairHom_right_comp_reflexion : right.comp reflexion = right_reflexion := rfl
+
+namespace WalkingParallelPair
+
+/-- The forgetful functor forgetting the common section -/
+def inclusionWalkingReflexivePair : WalkingParallelPair ⥤ WalkingReflexivePair :=
+  { obj := fun x => match x with
+      | one => WalkingReflexivePair.zero
+      | zero => WalkingReflexivePair.one
+    map := fun f => match f with
+      | WalkingParallelPairHom.left => left
+      | WalkingParallelPairHom.right => right
+      | WalkingParallelPairHom.id _ => WalkingReflexivePairHom.id _
+    map_comp := by
+      intro _ _ _ f g; cases f <;> cases g <;> rfl }
+
+variable {C : Type u} [Category.{v} C]
+
+@[simp]
+lemma whisker_inclusionWalkingReflexivePair_zero (F : WalkingReflexivePair ⥤ C) :
+    F.obj (inclusionWalkingReflexivePair.obj WalkingParallelPair.one) =
+      F.obj WalkingReflexivePair.zero :=
+  rfl
+
+@[simp]
+lemma whisker_inclusionWalkingReflexivePair_one (F : WalkingReflexivePair ⥤ C) :
+    F.obj (inclusionWalkingReflexivePair.obj WalkingParallelPair.zero) =
+      F.obj WalkingReflexivePair.one :=
+  rfl
+
+@[simp]
+lemma whisker_inclusionWalkingReflexivePair_left (F : WalkingReflexivePair ⥤ C) :
+    F.map (inclusionWalkingReflexivePair.map WalkingParallelPairHom.left) = F.map left := rfl
+
+@[simp]
+lemma whisker_inclusionWalkingReflexivePair_right (F : WalkingReflexivePair ⥤ C) :
+    F.map (inclusionWalkingReflexivePair.map WalkingParallelPairHom.right) = F.map right := rfl
+
+/-- The forgetful functor is a final functor -/
+instance inclusionWalkingReflexivePair_final : Functor.Final inclusionWalkingReflexivePair := by
+  constructor
+  set e₀ : (StructuredArrow WalkingReflexivePair.one inclusionWalkingReflexivePair) :=
+    StructuredArrow.mk (Y := WalkingParallelPair.zero) (id WalkingReflexivePair.one)
+  set e₁ : (StructuredArrow WalkingReflexivePair.zero inclusionWalkingReflexivePair) :=
+    StructuredArrow.mk (Y := WalkingParallelPair.one) (id WalkingReflexivePair.zero)
+  intro x
+  have h : Inhabited (StructuredArrow x inclusionWalkingReflexivePair) := by
+    · constructor
+      cases x with
+      | one => exact e₀
+      | zero => exact e₁
+  cases x with
+    | zero => apply IsConnected.of_induct (j₀ := e₁)
+              intro p h₁ h₂ t
+              rcases t with ⟨l, y, f⟩
```

Comment:
`intro/rcases` can be combined as:
```suggestion
              rintro p h₁ h₂ ⟨l, y, f⟩
```
(There is another similar situation below.)